catalyst/dl/runner/gan.py

```diff
@@ -61,18 +96,190 @@ def forward(self, batch):
+    ) -> OrderedDict:
+        """
+            Catalyst can't take several optimizer callbacks for generator
+            and discriminator and process them raw
+            Therefore, phase manager, criterion callbacks and wrappers
+            will be created to manage training order
+        """
+        assert isinstance(
+            callbacks, dict
+        ), "callbacks should be of dict[str, Callback] type"
+        discriminator_phase = self.TRAIN_PARAMS["discriminator_train_phase"]
+        generator_phase = self.TRAIN_PARAMS["generator_train_phase"]
+        callbacks["phase_manager"] = PhaseManagerCallback(
+            train_phases=OrderedDict(
+                [
+                    (discriminator_phase, discriminator_phase_num),
+                    (generator_phase, generator_phase_num),
+                ]
+            ),
+            valid_mode="all",
+        )
+        callbacks["loss_g"] = PhaseBatchWrapperCallback(
+            base_callback=CriterionCallback(
+                input_key="real_targets",
+                output_key="fake_logits",
+                prefix="loss_g",
+            ),
+            active_phases=[generator_phase],
+        )
+        callbacks["loss_d_real"] = PhaseBatchWrapperCallback(
+            base_callback=CriterionCallback(
+                input_key="real_targets",
+                output_key="real_logits",
+                prefix="loss_d_real",
+            ),
+            active_phases=[discriminator_phase],
+        )
+        callbacks["loss_d_fake"] = PhaseBatchWrapperCallback(
+            base_callback=CriterionCallback(
+                input_key="fake_targets",
+                output_key="fake_logits",
+                prefix="loss_d_fake",
+            ),
+            active_phases=[discriminator_phase],
+        )
+        callbacks["loss_d"] = PhaseBatchWrapperCallback(
+            base_callback=CriterionAggregatorCallback(
+                loss_keys=["loss_d_real", "loss_d_fake"],
+                loss_aggregate_fn="mean",
+                prefix="loss_d",
+            ),
+            active_phases=[discriminator_phase],
+        )
+        callbacks["optim_g"] = PhaseBatchWrapperCallback(
+            base_callback=OptimizerCallback(
+                loss_key="loss_g", optimizer_key=self.generator_key,
+            ),
+            active_phases=[generator_phase],
+        )
+        callbacks["optim_d"] = PhaseBatchWrapperCallback(
+            base_callback=OptimizerCallback(
+                loss_key="loss_d", optimizer_key=self.discriminator_key,
+            ),
+            active_phases=[discriminator_phase],
+        )
+        return OrderedDict(callbacks)
+
+    def train(
+        self,
+        models: Model,
+        criterion: Criterion,
+        optimizers: Dict[str, Optimizer],
+        datasets: Dict[str, Dataset],
+        loaders: Dict[str, DataLoader],
+        callbacks: Dict[str, Callback] = {},
+        discriminator_phase_num: int = 1,
```

Comment:
I believe it will be a better idea to require (or add the default) `PhaseCallback`. Let me explain this: there may be different phase manipulation strategies, e.g. train generator until discriminator is at most 75% right, train discriminator until it is at least 90% right. So in order to support more generic phase manipulation I propose to set directly PhaseCallback instead